File: app/scripts/model/graphUtils.ts
import {
  IEdgeData,
  IGraphData,
  IModelInfoReqs,
  IVertexData,
} from "../interfaces.js";
import { Layer } from "./layers/layers.js";

export interface IEdgesByVertex {
  [key: string]: {
    in: string[];
    out: string[];
  };
}

export class GraphUtils {
  public static edgesBetweenVertices(args: {
    graphData: IGraphData;
    edgesByVertex: IEdgesByVertex;
    vtxIds: string[];
  }): IModelInfoReqs["edgesBetweenVertices"]["response"] {
    const nonexistentVertices: string[] = [];
    for (const vtxId of args.vtxIds) {
      if (args.graphData.vertices[vtxId] === undefined) {
        nonexistentVertices.push(vtxId);
      }
    }
    if (nonexistentVertices.length !== 0) {
      return {
        requestNonexistentVertices: nonexistentVertices,
        verticesExist: false,
      };
    }

    const vtxOutputEdges = new Set<string>();
    const vtxInputEdges = new Set<string>();

    for (const vtxId of args.vtxIds) {
      for (const outEdgeId of args.edgesByVertex[vtxId].out) {
        vtxOutputEdges.add(outEdgeId);
      }
      for (const inEdgeId of args.edgesByVertex[vtxId].in) {
        vtxInputEdges.add(inEdgeId);
      }
    }

    const edgesBetween: {[key: string]: IEdgeData} = {};

    for (const edgeId of vtxOutputEdges) {
      if (vtxInputEdges.has(edgeId)) {
        edgesBetween[edgeId] = args.graphData.edges[edgeId];
      }
    }

    return {
      edges: edgesBetween,
      verticesExist: true,
    };
  }

  public static getUniqueEdgeIds(args: {
    graphData: IGraphData;
    count: number;
  }): IModelInfoReqs["getUniqueEdgeIds"]["response"] {
    const ids = new Set<string>();

    while (ids.size < args.count) {
      const randomVal = Math.random();
      let multiple = 10;
      let id: string;
      while (
        args.graphData.edges[id = Math.floor(randomVal * multiple).toString()] !== undefined ||
        ids.has(id)
      ) {
        multiple *= 100;
      }
      ids.add(id);
    }

    return {
      edgeIds: Array.from(ids),
    };
  }

  public static getUniqueVertexIds(args: {
    graphData: IGraphData;
    count: number;
  }): IModelInfoReqs["getUniqueVertexIds"]["response"] {
    const ids = new Set<string>();

    while (ids.size < args.count) {
      const randomVal = Math.random();
      let multiple = 100;
      let id: string;
      while (
        args.graphData.vertices[id = Math.floor(randomVal * multiple).toString()] !== undefined ||
        ids.has(id)
      ) {
        multiple *= 10;
      }
      ids.add(id);
    }

    return {
      vertexIds: Array.from(ids),
    };
  }

  public static validateMoveVertex(args: {
    graphData: IGraphData;
    vtxId: string;
    x: number;
    y: number;
  }): null | string {
    if (args.graphData.vertices[args.vtxId] === undefined) {
      return `Vertex with id ${args.vtxId} does not exist`;
    }

    return null;
  }

  public static moveVertex(args: {
    graphData: IGraphData;
    vtxId: string;
    x: number;
    y: number;
  }): void {
    const vtx = args.graphData.vertices[args.vtxId];
    if (vtx === undefined) {
      throw new Error(`Could not find vertex with id ${args.vtxId}`);
    }

    vtx.geo.x = args.x;
    vtx.geo.y = args.y;
  }

  public static validateDeleteVertex(args: {
    graphData: IGraphData;
    edgesByVertex: IEdgesByVertex;
    vertexId: string;
  }): string | null {
    if (args.graphData.vertices[args.vertexId] === undefined) {
      return `Could not find vertex with id ${args.vertexId}`;
    }

    return null;
  }

  public static deleteVertex(args: {
    graphData: IGraphData;
    edgesByVertex: IEdgesByVertex;
    vertexId: string;
  }): void {
    if (args.graphData.vertices[args.vertexId] === undefined) {
      throw new Error(`Could not find vertex with id ${args.vertexId}`);
    }

    const connectedEdges: string[] = [];
    for (const edgeId in args.graphData.edges) {
      if (
        args.graphData.edges[edgeId].sourceVertexId === args.vertexId ||
        args.graphData.edges[edgeId].targetVertexId === args.vertexId
      ) {
        connectedEdges.push(edgeId);
      }
    }
    for (const connectedEdge of connectedEdges) {
      this.deleteEdge({
        edgeId: connectedEdge,
        edgesByVertex: args.edgesByVertex,
        graphData: args.graphData,
      });
    }

    delete args.graphData.vertices[args.vertexId];
    delete args.edgesByVertex[args.vertexId];
  }

  public static deleteEdge(args: {
    graphData: IGraphData;
    edgesByVertex: IEdgesByVertex;
    edgeId: string;
  }): void {
    if (args.graphData.edges[args.edgeId] === undefined) {
      throw new Error(`Could not find edge with id ${args.edgeId}`);
    }

    const edge = args.graphData.edges[args.edgeId];
    const inIdx = args.edgesByVertex[edge.sourceVertexId].out.indexOf(args.edgeId);
    args.edgesByVertex[edge.sourceVertexId].out.splice(inIdx, 1);
    const outIdx = args.edgesByVertex[edge.targetVertexId].in.indexOf(args.edgeId);
    args.edgesByVertex[edge.targetVertexId].in.splice(outIdx, 1);

    delete args.graphData.edges[args.edgeId];
  }

  public static validateDeleteEdge(args: {
    graphData: IGraphData;
    edgesByVertex: IEdgesByVertex;
    edgeId: string;
  }): string | null {
    if (args.graphData.edges[args.edgeId] === undefined) {
      return `Could not find edge with id ${args.edgeId}`;
    }

    return null;
  }

  public static validateCloneVertex(args: {
    graphData: IGraphData;
    edgesByVertex: IEdgesByVertex;
    newVtxId: string;
    oldVtxId: string;
    x: number;
    y: number;
  }): string | null {
    if (args.graphData.vertices[args.oldVtxId] === undefined) {
      return `Vertex with id ${args.oldVtxId} does not exist`;
    }

    if (args.graphData.vertices[args.newVtxId] !== undefined) {
      return `Vertex with id ${args.newVtxId} already exists`;
    }

    const oldVtx = args.graphData.vertices[args.oldVtxId];
    if (oldVtx === undefined) {
      return `Could not find vertex with id ${args.oldVtxId}`;
    }

    return null;
  }

  public static cloneVertex(args: {
    graphData: IGraphData;
    edgesByVertex: IEdgesByVertex;
    newVtxId: string;
    oldVtxId: string;
    x: number;
    y: number;
  }): void {
    if (args.graphData.vertices[args.newVtxId] !== undefined) {
      throw new Error(`Vertex with id ${args.newVtxId} already exists`);
    }

    const oldVtx = args.graphData.vertices[args.oldVtxId];
    if (oldVtx === undefined) {
      throw new Error(`Coudl not find vertex with id ${args.oldVtxId}`);
    }

    const newVtx: IVertexData = JSON.parse(JSON.stringify(oldVtx));
    newVtx.geo.x = args.x;
    newVtx.geo.y = args.y;

    this.addVertex({
      edgesByVertex: args.edgesByVertex,
      graphData: args.graphData,
      id: args.newVtxId,
      vtxData: newVtx,
    });
  }

  public static createVertexFromLayer(args: {
    layer: Layer;
    x?: number;
    y?: number;
  }): IVertexData {
    const vtxData: IVertexData = {
      geo: {
        x: args.x === undefined ? 0 : args.x,
        y: args.y === undefined ? 0 : args.y,
      },
      label: args.layer.getType(),
      ports: {},
    };
    const inputPortCount = args.layer.getPortIds().filter((id) => args.layer.getPortInfo(id).type === "input").length;
    const outputPortCount = args.layer.getPortIds().filter((id) => args.layer.getPortInfo(id).type === "output").length;

    let inputPortIdx = 0;
    let outputPortIdx = 0;
    for (const portId of args.layer.getPortIds()) {
      const portInfo = args.layer.getPortInfo(portId);
      vtxData.ports[portId] = {
        portType: portInfo.type,
        position: portInfo.type === "input" ? (
          1 / (inputPortCount + 1) * ++inputPortIdx
        ) : (
          1 / (outputPortCount + 1) * ++outputPortIdx
        ),
        side: portInfo.type === "input" ? "top" : "bottom",
      };
    }
    return vtxData;
  }

  public static addVertex(args: {
    graphData: IGraphData;
    edgesByVertex: IEdgesByVertex;
    id: string;
    vtxData: IVertexData;
  }): void {
    args.graphData.vertices[args.id] = args.vtxData;
    args.edgesByVertex[args.id] = {in: [], out: []};
  }

  public static createEdge(args: {
    graphData: IGraphData;
    edgesByVertex: IEdgesByVertex;
    newEdgeId: string;
    sourceVtxId: string;
    sourcePortId: string;
    targetVtxId: string;
    targetPortId: string;
  }): void {
    const edgeValidationMessage = this.validateCreateEdge(args);

    if (edgeValidationMessage !== null) {
      throw new Error(`Invalid edge: ${edgeValidationMessage}`);
    }

    const edge: IEdgeData = {
      consistency: "consistent",
      sourcePortId: args.sourcePortId,
      sourceVertexId: args.sourceVtxId,
      targetPortId: args.targetPortId,
      targetVertexId: args.targetVtxId,
    };
    args.graphData.edges[args.newEdgeId] = edge;
    args.edgesByVertex[edge.sourceVertexId].out.push(args.newEdgeId);
    args.edgesByVertex[edge.targetVertexId].in.push(args.newEdgeId);
  }

  public static validateCreateEdge(args: {
    graphData: IGraphData;
    edgesByVertex: IEdgesByVertex;
    newEdgeId: string;
    sourceVtxId: string;
    sourcePortId: string;
    targetVtxId: string;
    targetPortId: string;
  }): string | null {
    if (args.graphData.edges[args.newEdgeId] !== undefined) {
      return `Edge with id ${args.newEdgeId} already exists`;
    }

    const sourceVertex = args.graphData.vertices[args.sourceVtxId];
    const targetVertex = args.graphData.vertices[args.targetVtxId];
    if (sourceVertex === undefined) {
      return "Source does not exist";
    }
    if (targetVertex === undefined) {
      return "Target does not exist";
    }

    const sourcePort = sourceVertex.ports[args.sourcePortId];
    const targetPort = targetVertex.ports[args.targetPortId];

    if (sourcePort === undefined) {
      return "Source port does not exist";
    }
    if (targetPort === undefined) {
      return "Target port does not exist";
    }

    if (sourcePort.portType !== "output") {
      return "Source is not an output port";
    }
    if (targetPort.portType !== "input") {
      return "Target is not an input port";
    }

    // check that there isn't an identical edge present
    const sourceOutEdges = args.edgesByVertex[args.sourceVtxId].out;
    const targetInEdges = args.edgesByVertex[args.targetVtxId].in;
    const edgesFromSourceToTarget = sourceOutEdges.filter((edgeId) => targetInEdges.indexOf(edgeId) !== -1);
    for (const edgeId of edgesFromSourceToTarget) {
      const edge = args.graphData.edges[edgeId];

      // check if the edge is identical
      if (edge.sourcePortId === args.sourcePortId && edge.targetPortId === args.targetPortId) {
        return "Identical connection already exists";
      }
    }

    // Loop detection
    const sourceAncestorIds: string[] = [];

    const vertexIdsToInvestigate: string[] = [args.sourceVtxId];
    while (vertexIdsToInvestigate.length > 0) {
      const vtxIdToInvestigate = vertexIdsToInvestigate.pop() as string;
      sourceAncestorIds.push(vtxIdToInvestigate);

      for (const edgeId of args.edgesByVertex[vtxIdToInvestigate].in) {
        const edgeData = args.graphData.edges[edgeId];
        // check if the edge's source has not been seen before
        if (
          vertexIdsToInvestigate.indexOf(edgeData.sourceVertexId) === -1 &&
          sourceAncestorIds.indexOf(edgeData.sourceVertexId) === -1
        ) {
          vertexIdsToInvestigate.push(edgeData.sourceVertexId);
        }
      }
    }

    // return true if the target vertex is not an ancestor of the source vertex
    if (sourceAncestorIds.indexOf(args.targetVtxId) === -1) {
      return null;
    } else {
      return "Loop detected";
    }
  }
}

File: cs_graph/model/networkGraph/graphUtils.cs
using System.Collections.Generic;
using System.Linq;

namespace GraphUtilsNS {
  public static class GraphUtils {
    public static ResponseJson.GraphData getResponseJsonData(NetworkContainersNS.Graph graph) {
      ResponseJson.GraphData responseJsonData = new ResponseJson.GraphData {
        vertices = new Dictionary<string, ResponseJson.VertexData>(),
        edges = new Dictionary<string, ResponseJson.EdgeData>()
      };

      foreach (KeyValuePair<string, NetworkContainersNS.Vertex> vtxEntry in graph.vertices) {
        responseJsonData.vertices[vtxEntry.Key] = GraphUtils.getVertexJsonData(vtxEntry.Value);
      }

      foreach(KeyValuePair<string, NetworkContainersNS.Edge> edgeEntry in graph.edges) {
        responseJsonData.edges[edgeEntry.Key] = GraphUtils.getEdgeJsonData(edgeEntry.Value);
      }

      return responseJsonData;
    }

    private static ResponseJson.VertexData getVertexJsonData(NetworkContainersNS.Vertex vtx) {
      ResponseJson.VertexData jsonVtxData = new ResponseJson.VertexData {
        label = vtx.label,
        geo = new ResponseJson.GeoData {
          x = vtx.xLocation,
          y = vtx.yLocation,
        },
        ports = new Dictionary<string, ResponseJson.GraphPortData>()
      };

      foreach (KeyValuePair<string, NetworkContainersNS.NetworkPort> edgeEntry in vtx.ports) {
        jsonVtxData.ports[edgeEntry.Key] = new ResponseJson.GraphPortData {
          side = GraphUtils.getPortSideString(edgeEntry.Value.side),
          position = edgeEntry.Value.position,
          portType = GraphUtils.getPortTypeString(edgeEntry.Value.type)
        };
      }

      return jsonVtxData;
    }

    private static string getPortTypeString(NetworkContainersNS.PortType portType) {
      switch (portType) {
      case NetworkContainersNS.PortType.Input: return "input";
      case NetworkContainersNS.PortType.Output: return "output";
      default: throw new System.ArgumentOutOfRangeException();
      }
    }

    private static string getPortSideString(NetworkContainersNS.SideType sideType) {
      switch (sideType) {
      case NetworkContainersNS.SideType.Top: return "top";
      case NetworkContainersNS.SideType.Bottom: return "bottom";
      case NetworkContainersNS.SideType.Left: return "left";
      case NetworkContainersNS.SideType.Right: return "right";
      default: throw new System.ArgumentOutOfRangeException();
      }
    }

    private static string getEdgeConsistencyString(NetworkContainersNS.ConsistencyType consistencyType) {
      switch (consistencyType) {
      case NetworkContainersNS.ConsistencyType.Consistent: return "consistent";
      case NetworkContainersNS.ConsistencyType.Inconsistent: return "inconsistent";
      default: throw new System.ArgumentOutOfRangeException();
      }
    }

    public static ResponseJson.EdgeData getEdgeJsonData(NetworkContainersNS.Edge edge) {
      return new ResponseJson.EdgeData {
        consistency = GraphUtils.getEdgeConsistencyString(edge.consistency),
        sourceVertexId = edge.sourceVertexId,
        sourcePortId = edge.sourcePortId,
        targetVertexId = edge.targetVertexId,
        targetPortId = edge.targetPortId
      };
    }

    public static void moveVertex(
      NetworkContainersNS.Graph graph,
      string vertexId,
      float x,
      float y
    ) {
      NetworkContainersNS.Vertex vtx = graph.vertices[vertexId];
      vtx.xLocation = x;
      vtx.yLocation = y;
    }

    public static string validateEdge(
      NetworkContainersNS.Graph graph,
      Dictionary<string, ModelClassNS.VertexEdgesInfo> edgesByVertex,
      string newEdgeId,
      string sourceVertexId,
      string sourcePortId,
      string targetVertexId,
      string targetPortId
    ) {
      // check vertices and ports
      if (!graph.vertices.ContainsKey(sourceVertexId)) {
        return "Vertex with given source id does not exist";
      }
      if (!graph.vertices.ContainsKey(targetVertexId)) {
        return "Vertex with given target id does not exist";
      }
      if (!graph.vertices[sourceVertexId].ports.ContainsKey(sourcePortId)) {
        return "Source vertex does not have port with given id";
      }
      if (!graph.vertices[targetVertexId].ports.ContainsKey(targetPortId)) {
        return "Target vertex does not have port with given id";
      }
      if (graph.vertices[sourceVertexId].ports[sourcePortId].type != NetworkContainersNS.PortType.Output) {
        return "Source port is not an output port";
      }
      if (graph.vertices[targetVertexId].ports[targetPortId].type != NetworkContainersNS.PortType.Input) {
        return "Target port is not an input port";
      }

      // check if an identical edge already exists
      var edgesBetweenTwoVtxs = edgesByVertex[sourceVertexId].edgesOut.Where(id => edgesByVertex[targetVertexId].edgesIn.Contains(id));
      foreach (string edgeId in edgesBetweenTwoVtxs) {
        if (
          graph.edges[edgeId].sourcePortId == sourcePortId &&
          graph.edges[edgeId].targetPortId == targetPortId
        ) {
          return "Identical connection already exists";
        }
      }

      // search for loop
      HashSet<string> vertexIdsToInvestigate = new HashSet<string>() {sourceVertexId};
      HashSet<string> ancestorsOfSourceVtx = new HashSet<string>();

      while (vertexIdsToInvestigate.Count != 0) {
        string investigateVertexId = vertexIdsToInvestigate.ElementAt(0);
        vertexIdsToInvestigate.Remove(investigateVertexId);
        ancestorsOfSourceVtx.Add(investigateVertexId);

        foreach (string edgeInId in edgesByVertex[investigateVertexId].edgesIn) {
          string edgeSourceVertex = graph.edges[edgeInId].sourceVertexId;

          if (
            !vertexIdsToInvestigate.Contains(edgeSourceVertex) &&
            !ancestorsOfSourceVtx.Contains(edgeSourceVertex)
          ) {
            vertexIdsToInvestigate.Add(edgeSourceVertex);
          }
        }
      }

      if (ancestorsOfSourceVtx.Contains(targetVertexId)) {
        return "Loop detected";
      }

      return null;
    }

    public static void deleteVertex(
      NetworkContainersNS.Graph graph,
      Dictionary<string, ModelClassNS.VertexEdgesInfo> edgesByVertex,
      string vertexId
    ) {
      if (!graph.vertices.ContainsKey(vertexId)) {
        throw new System.Exception("Vertex with given id does not exist");
      }

      // Use ToList because C# won't allow the edges by vertex collection to be modified while it is being iterated over
      foreach (string edgeId in edgesByVertex[vertexId].edgesIn.Union(edgesByVertex[vertexId].edgesOut).ToList()) {
        GraphUtils.deleteEdge(graph, edgesByVertex, edgeId);
      }

      graph.vertices.Remove(vertexId);
      edgesByVertex.Remove(vertexId);
    }

    public static void deleteEdge(
      NetworkContainersNS.Graph graph,
      Dictionary<string, ModelClassNS.VertexEdgesInfo> edgesByVertex,
      string edgeId
    ) {
      if (!graph.edges.ContainsKey(edgeId)) {
        throw new System.Exception("Edge with given id does not exist");
      }

      NetworkContainersNS.Edge edge = graph.edges[edgeId];
      graph.edges.Remove(edgeId);

      edgesByVertex[edge.sourceVertexId].edgesOut.Remove(edgeId);
      edgesByVertex[edge.targetVertexId].edgesIn.Remove(edgeId);
    }

    public static void createEdge(
      NetworkContainersNS.Graph graph,
      Dictionary<string, ModelClassNS.VertexEdgesInfo> edgesByVertex,
      string newEdgeId,
      string sourceVertexId,
      string sourcePortId,
      string targetVertexId,
      string targetPortId
    ) {
      // nullable
      string validated = GraphUtils.validateEdge(
        graph,
        edgesByVertex,
        newEdgeId,
        sourceVertexId,
        sourcePortId,
        targetVertexId,
        targetPortId
      );

      if (validated != null) {
        throw new System.Exception(validated);
      }

      graph.edges[newEdgeId] = new NetworkContainersNS.Edge {
        consistency =  NetworkContainersNS.ConsistencyType.Consistent,
        sourceVertexId = sourceVertexId,
        sourcePortId = sourcePortId,
        targetVertexId = targetVertexId,
        targetPortId = targetPortId,
      };

      edgesByVertex[sourceVertexId].edgesOut.Add(newEdgeId);
      edgesByVertex[targetVertexId].edgesIn.Add(newEdgeId);
    }

    public static void cloneVertex(
      NetworkContainersNS.Graph graph,
      Dictionary<string, ModelClassNS.VertexEdgesInfo> edgesByVertex,
      string sourceVertexId,
      string newVertexId,
      float x,
      float y
    ) {
      if (graph.vertices.ContainsKey(newVertexId)) {
        throw new System.Exception("Vertex with new key already exists");
      }
      if (!graph.vertices.ContainsKey(sourceVertexId)) {
        throw new System.Exception("Vertex with source id does not exist");
      }
      
      NetworkContainersNS.Vertex newVtx = graph.vertices[sourceVertexId].clone();
      graph.vertices[newVertexId] = newVtx;
      newVtx.xLocation = x;
      newVtx.yLocation = y;

      edgesByVertex[newVertexId] = new ModelClassNS.VertexEdgesInfo {
        edgesIn = new List<string>(),
        edgesOut = new List<string>()
      };
    }

    private static List<string> getUniqueIds(
      System.Func<string, bool> testFunc,
      int count
    ) {
      var random = new System.Random();
      HashSet<string> uniqueIds = new HashSet<string>();

      for (int i = 0; i < count; i++) {
        // const float random
        double dbl = random.NextDouble();
        int multiplier = 10;
        bool done = false;
        string id = "";
        while (!done) {
          id = System.Convert.ToInt32(dbl*multiplier).ToString();
          done = (!uniqueIds.Contains(id)) && (!testFunc(id));
          multiplier *= 10;
        }
        uniqueIds.Add(id);
      }

      return new List<string>(uniqueIds);      
    }

    public static List<string> getUniqueVertexIds(
      NetworkContainersNS.Graph graph,
      int count
    ) {
      return GraphUtils.getUniqueIds(
        (string id) => {
          return graph.vertices.ContainsKey(id);
        },
        count
      );
    }

    public static List<string> getUniqueEdgeIds(
      NetworkContainersNS.Graph graph,
      int count
    ) {
      return GraphUtils.getUniqueIds(
        (string id) => {
          return graph.edges.ContainsKey(id);
        },
        count
      );
    }

    public static List<string> getEdgesBetweenVertices(
      NetworkContainersNS.Graph graph,
      Dictionary<string, ModelClassNS.VertexEdgesInfo> edgesByVertex,
      List<string> vertexIds
    ) {
      var edgesOut = new HashSet<string>();
      var edgesIn = new HashSet<string>();
      
      foreach (string vtxId in vertexIds) {
        if (!edgesByVertex.ContainsKey(vtxId)) {
          throw new System.Exception("No such vertex id");
        }
        
        edgesByVertex[vtxId].edgesIn.ForEach((string edgeId) => edgesIn.Add(edgeId));
        edgesByVertex[vtxId].edgesOut.ForEach((string edgeId) => edgesOut.Add(edgeId));
      }

      edgesOut.IntersectWith(edgesIn);

      return new List<string>(edgesOut);
    }
  }
}
